### ISSUE TITLE
An IfElse arm that did not run can poison a parameter's gradient with NaN

### ISSUE THREAD

**Issue by Eric-Louchez** [Shorokoo/Shorokoo#313]
## Summary

An `IfElse` arm's forward is computed whether or not the condition picks it, wherever the backward
pass reads one of its values: reverse-mode autodiff needs the forward's intermediates, and a value
read unconditionally cannot be computed conditionally. The arm that did not run is handed a zero
incoming gradient, so its contribution is mathematically zero — and arithmetic usually delivers
that.

Not always. An arm guarding a computation that is invalid off its own path has a non-finite
derivative there, and NaN times zero is NaN, not zero. That NaN lands in the parameter's gradient
and destroys the weight on the very step the arm was *not* taken.

## Repro

```csharp
[Module]
public partial class SqrtInOneIfArmModel
{
    public static Tensor<float32> Inline(Tensor<float32> t, Scalar<bit> cond)
    {
        var w = Ones.Init([Scalar(2L)]);
        return cond.IfElse((t * w).Sqrt(), t * w);
    }
}
```

Train one step on `t = [-1, -4]` with `cond = false`, so the `else` arm runs and `sqrt` is never
selected:

```
loss = 8.5          ← right: the else arm ran
w    = [NaN, NaN]   ← the arm that did not run took the weights with it
```

`sqrt` of a negative is NaN, its derivative is NaN, and the zero gradient routed to the untaken arm
cannot clear it.

## Fix

Selecting is not arithmetic: an `If` hands back the branch it picks and never reads the other. Where
a gradient crosses out of an arm it is now selected against the arm's own condition, with zeros of
its shape on the other side — so the arm's chain may carry a NaN and none of it reaches a parameter.
The gate sits at the boundary rather than at every step, since only what leaves an arm can reach
one. An arm is read off what the `IF_CLOSE`'s branch inputs reach exclusively; the inputs and
parameters a branch merely reads are nobody's arm, which is what makes a gradient heading for one a
gradient *leaving* the arm.

After the fix the repro trains to `w = [0.9, -0.6]` — the same weights as the same input with the
other arm finite, which is the point.

## Correction to this issue as first filed

It originally claimed that a training graph runs both `IfElse` arms outright. That was wrong, and
measured wrongly: `TrainingStepPureGraph` is the graph *before* `ConfigureScopes`, so its `If`
bodies are empty by construction. On the compiled trainstep the arms keep everything except the
forward values the backward reads — 3 of 5 nodes in one arm of the repro used there. That residue is
cost, not correctness, and is recorded in `Documentation/limitations.md` rather than tracked here.

The NaN above is pre-existing and not a consequence of #270 or #312: it reproduces identically on
`main`, where an eager `IfElse` is never scoped at all.

## Pinned

`TrainingRigTrainingLoopCoverageTests.TestTheIfElseArmThatDidNotRunLeavesTheGradientAlone` trains the
same input twice — once with the other arm finite, once not — and asserts the same weights.
